File: apps/ios/Sources/Models/SavedSessionsStore.swift
import Foundation
import Observation

/// Lifecycle bucket mirrored from `core/src/saved/mod.rs::SavedSessionStatus`.
/// Stringly-encoded `lowercase` to match the Rust serde repr —
/// `#[serde(rename_all = "lowercase")]` on the Rust enum.
enum SavedSessionStatus: String, Codable, Equatable {
    case live
    case exited
    case unknown
}

/// Swift mirror of `core/src/saved/mod.rs::SavedSession`. The on-disk
/// JSON is shared with the Rust core (no UniFFI surface yet — the
/// "Sessions" screen reads + writes this file directly so we can
/// iterate on the shape on a single platform before promoting it).
///
/// `CodingKeys` map Swift's camelCase fields to the Rust snake_case
/// serde repr so the same JSON is round-trippable from either side.
struct SavedSession: Codable, Equatable, Identifiable {
    var id: String
    var serverID: String
    var agent: String
    var cwd: String?
    var firstSeen: String
    var lastSeen: String
    var messageCount: UInt32
    var summary: String
    var status: SavedSessionStatus

    /// Compound identity for `SwiftUI.ForEach` — `id` alone is not
    /// unique across servers (the harness mints UUIDs per session, but
    /// nothing prevents a collision if a user paired two harnesses that
    /// happen to reuse the same id). Match the Rust `compound_key`.
    var compoundID: String { "\(serverID)::\(id)" }

    enum CodingKeys: String, CodingKey {
        case id
        case serverID = "server_id"
        case agent
        case cwd
        case firstSeen = "first_seen"
        case lastSeen = "last_seen"
        case messageCount = "message_count"
        case summary
        case status
    }
}

/// Persisted index of every session the client has ever seen, across
/// every server. Backs the "Resume an old thread" screen (upstream parity
/// audit item A.8). Mirrors `core/src/saved/mod.rs::SavedSessionStore`:
///
/// * Persistence path: `Application Support/conduit/saved-sessions.json`.
/// * Wire shape: `{"sessions": {"<server_id>::<session_id>": SavedSession, ...}}`,
///   matching the Rust `HashMap` serde repr so either side can produce
///   or consume the file without a schema dance.
/// * Singleton: there's only one historical index per device. We expose
///   `shared` so call sites (`SessionStore.ingestStatus`, the
///   `SessionsScreen` view) don't have to thread an instance through.
@Observable
@MainActor
final class SavedSessionsStore {

    /// Process-wide singleton. The first access reads the JSON file off
    /// disk synchronously — the file is tiny (one record per session
    /// the user has ever seen) so blocking on init is fine and saves
    /// the screen from a "loading…" flash.
    static let shared = SavedSessionsStore()

    /// All known rows. Mutations go through `upsert`, which keeps this
    /// in sync with disk; reads in the UI go through `recent`.
    private(set) var sessions: [SavedSession] = []

    /// Persisted set of session ids the user has explicitly deleted.
    /// A tombstoned id is permanently suppressed from `upsert` (so a
    /// status/list refresh can't re-add it) and excluded from `recent`
    /// (so the history screen never shows it). Keyed by bare session id
    /// — the harness mints UUIDs unique per session, and a delete is
    /// terminal across servers, matching `remove(id:)`. We cap the set
    /// at `tombstoneCap` newest-first so it can't grow unbounded; an
    /// id evicted from the cap can theoretically reappear, but only
    /// after `tombstoneCap` *other* sessions have been deleted, by
    /// which point the broker has long since reaped the original.
    private(set) var deletedIDs: Set<String> = []

    /// Insertion-ordered companion to `deletedIDs` so we can evict the
    /// oldest tombstone when the cap is exceeded. Newest at the end.
    private var deletedOrder: [String] = []

    /// Upper bound on retained tombstones. Generous — the on-disk cost
    /// is one short string per entry.
    private let tombstoneCap = 500

    /// Resolved persistence path, exposed for tests + diagnostics.
    let storeURL: URL

    init(storeURL: URL? = nil) {
        self.storeURL = storeURL ?? SavedSessionsStore.defaultStoreURL()
        let loaded = Self.loadFromDisk(at: self.storeURL)
        self.sessions = loaded.sessions
        self.deletedOrder = loaded.deletedOrder
        self.deletedIDs = Set(loaded.deletedOrder)
    }

    /// Latest-first slice clamped to `limit`. Ties broken by id so
    /// snapshot tests stay deterministic — mirrors the Rust
    /// `list_recent` ordering rule.
    func recent(limit: Int = 200) -> [SavedSession] {
        // Belt-and-braces: a tombstoned row is normally already absent
        // from `sessions` (delete removes it), but filter here too so a
        // race (status frame in flight during delete) can never leak a
        // deleted session into the history screen.
        let sorted = sessions
            .filter { !deletedIDs.contains($0.id) }
            .sorted { lhs, rhs in
                if lhs.lastSeen != rhs.lastSeen { return lhs.lastSeen > rhs.lastSeen }
                return lhs.id < rhs.id
            }
        if sorted.count <= limit { return sorted }
        return Array(sorted.prefix(limit))
    }

    /// True when the id has been explicitly deleted by the user. Call
    /// sites (e.g. `SessionStore.refreshSessions`) consult this to keep
    /// a tombstoned session out of the *live* list as well, since the
    /// broker can keep reporting a deleted session whose tmux lingers.
    func isTombstoned(id: String) -> Bool {
        deletedIDs.contains(id)
    }

    /// Fold a live `ProjectSession` + `SessionStatus` snapshot into the
    /// saved index. Idempotent — calling with the same input twice
    /// leaves the file unchanged. Mirrors the Rust `upsert` algorithm:
    /// first_seen is sticky, last_seen advances, the first non-empty
    /// summary wins, and once we record `.exited` the row is locked.
    func upsert(
        session: ProjectSession,
        serverID: String,
        status: SessionStatus?,
        firstUserMessage: String?,
        messageCount: Int,
        isExited: Bool
    ) {
        // A deleted session must STAY deleted. The deployed broker keeps
        // tmux-backed PTYs alive (#199), so a just-deleted session can
        // still surface on the next listSessions/status delta and feed
        // back into here. Without this guard it would be re-added and
        // reappear in history (reading as live → interactive). The
        // tombstone is the client-side guarantee.
        if deletedIDs.contains(session.id) { return }
        let now = status?.lastActivityAt
            ?? status?.startedAt
            ?? session.lastActivityAt
            ?? session.startedAt
            ?? ISO8601DateFormatter().string(from: Date())
        let summary = Self.truncateSummary(firstUserMessage ?? "")
        let nextStatus: SavedSessionStatus = isExited ? .exited : .live
        let agent = session.assistant
        let cwd = status?.cwd ?? session.cwd

        if let idx = sessions.firstIndex(where: {
            $0.id == session.id && $0.serverID == serverID
        }) {
            var row = sessions[idx]
            row.lastSeen = max(row.lastSeen, now)
            row.messageCount = max(row.messageCount, UInt32(clamping: messageCount))
            row.agent = agent
            if let cwd { row.cwd = cwd }
            if row.summary.isEmpty, !summary.isEmpty {
                row.summary = summary
            }
            row.status = mergeStatus(existing: row.status, next: nextStatus)
            // Avoid a no-op write — checked equality before persisting.
            if row != sessions[idx] {
                sessions[idx] = row
                persist()
            }
        } else {
            let row = SavedSession(
                id: session.id,
                serverID: serverID,
                agent: agent,
                cwd: cwd,
                firstSeen: now,
                lastSeen: now,
                messageCount: UInt32(clamping: messageCount),
                summary: summary,
                status: nextStatus
            )
            sessions.append(row)
            persist()
        }
    }

    /// Drop every row + tombstone. Test-only convenience — the
    /// production app has no "clear history" affordance yet.
    func reset() {
        sessions = []
        deletedIDs = []
        deletedOrder = []
        persist()
    }

    /// Tombstone + remove every saved-session row whose session-id
    /// equals `id`. Idempotent. Used by every delete path
    /// (`SessionStore.exit`): the live row is already gone, this records
    /// the tombstone so a subsequent `upsert` (from a status/list
    /// refresh while the broker's tmux lingers) can NEVER re-add it, and
    /// clears the persistent `Resume` entry so the row doesn't reappear
    /// on next launch. Matches across servers because the harness mints
    /// UUIDs unique per session and the user expects delete to be
    /// terminal.
    func remove(id: String) {
        let removedRow = sessions.contains { $0.id == id }
        sessions.removeAll { $0.id == id }
        let newTombstone = deletedIDs.insert(id).inserted
        if newTombstone {
            deletedOrder.append(id)
            // Cap the tombstone set so it can't grow forever. Evict the
            // oldest ids; by the time we've deleted `tombstoneCap`
            // sessions the broker has long reaped the early ones, so an
            // evicted tombstone is harmless.
            if deletedOrder.count > tombstoneCap {
                let overflow = deletedOrder.count - tombstoneCap
                let evicted = deletedOrder.prefix(overflow)
                deletedOrder.removeFirst(overflow)
                for id in evicted where !deletedOrder.contains(id) {
                    deletedIDs.remove(id)
                }
            }
        }
        if removedRow || newTombstone {
            persist()
        }
    }

    // MARK: - Persistence

    private func persist() {
        do {
            try Self.write(sessions: sessions, deletedOrder: deletedOrder, to: storeURL)
        } catch {
            // Persistence failure is logged but never propagated — the
            // saved store is best-effort. The in-memory state is still
            // accurate for the current session; we just won't survive
            // a relaunch. (Same posture as `SessionStore.persist`.)
            #if DEBUG
            print("SavedSessionsStore persist failed: \(error)")
            #endif
        }
    }

    nonisolated static func defaultStoreURL() -> URL {
        let base = (try? FileManager.default.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )) ?? FileManager.default.temporaryDirectory
        return base
            .appendingPathComponent("conduit", isDirectory: true)
            .appendingPathComponent("saved-sessions.json", isDirectory: false)
    }

    nonisolated private static func loadFromDisk(
        at url: URL
    ) -> (sessions: [SavedSession], deletedOrder: [String]) {
        guard let data = try? Data(contentsOf: url) else { return ([], []) }
        let decoder = JSONDecoder()
        if let envelope = try? decoder.decode(StoreEnvelope.self, from: data) {
            // De-dup the persisted tombstone list while preserving order.
            var seen = Set<String>()
            let order = (envelope.deletedIDs ?? []).filter { seen.insert($0).inserted }
            return (Array(envelope.sessions.values), order)
        }
        // Belt-and-braces: also accept a bare top-level array in case a
        // future Rust revision flattens the envelope.
        if let bare = try? decoder.decode([SavedSession].self, from: data) {
            return (bare, [])
        }
        return ([], [])
    }

    nonisolated private static func write(
        sessions: [SavedSession],
        deletedOrder: [String],
        to url: URL
    ) throws {
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        let envelope = StoreEnvelope(
            sessions: Dictionary(
                uniqueKeysWithValues: sessions.map { ($0.compoundID, $0) }
            ),
            deletedIDs: deletedOrder.isEmpty ? nil : deletedOrder
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(envelope)
        try data.write(to: url, options: .atomic)
    }

    /// Matches the Rust serde wrapper
    /// `struct SavedSessionStore { sessions: HashMap<String, SavedSession> }`,
    /// extended with an optional `deleted_ids` tombstone list. The field
    /// is optional + omitted-when-empty so the Rust core (which does not
    /// yet model it) round-trips the file untouched — serde ignores
    /// unknown fields by default, and an absent field decodes to nil
    /// here. Ordered newest-last for cap eviction.
    private struct StoreEnvelope: Codable {
        var sessions: [String: SavedSession]
        var deletedIDs: [String]?

        enum CodingKeys: String, CodingKey {
            case sessions
            case deletedIDs = "deleted_ids"
        }
    }

    // MARK: - Helpers

    /// UTF-8 safe truncation to 100 chars (`SUMMARY_MAX_CHARS` in Rust).
    /// Drops to the first line, trims, and appends `…` when the input is
    /// longer than the budget. Mirrors `core/src/saved/mod.rs::truncate_summary`
    /// so the Swift-side computed summary matches what the Rust side
    /// would have written for the same input.
    nonisolated static func truncateSummary(_ text: String) -> String {
        let firstLine = text
            .split(whereSeparator: { $0.isNewline })
            .first
            .map(String.init) ?? ""
        let cleaned = firstLine.trimmingCharacters(in: .whitespacesAndNewlines)
        let budget = 100
        if cleaned.count <= budget { return cleaned }
        return String(cleaned.prefix(budget - 1)) + "…"
    }

    private func mergeStatus(existing: SavedSessionStatus, next: SavedSessionStatus) -> SavedSessionStatus {
        // Exited is terminal — once we've seen it, never resurrect. Match
        // the Rust `merge_status` lattice (Unknown ⊏ Live ⊏ Exited).
        if existing == .exited || next == .exited { return .exited }
        if existing == .live || next == .live { return .live }
        return .unknown
    }
}

File: apps/android/app/src/main/kotlin/sh/nikhil/conduit/SavedSessionsStore.kt
package sh.nikhil.conduit

import org.json.JSONObject
import uniffi.conduit_core.ProjectSession
import uniffi.conduit_core.SessionStatus

/**
 * Lifecycle bucket for a persisted (archived) session row. Mirror of iOS
 * `SavedSessionStatus` / Rust `core/src/saved/mod.rs::SavedSessionStatus`.
 * `exited` is terminal — once recorded, never resurrected (see
 * [SavedSessionsReducer.mergeStatus]).
 */
enum class SavedSessionStatus(val raw: String) {
    LIVE("live"),
    EXITED("exited"),
    UNKNOWN("unknown");

    companion object {
        fun fromRaw(raw: String?): SavedSessionStatus = when (raw?.trim()?.lowercase()) {
            "live" -> LIVE
            "exited" -> EXITED
            else -> UNKNOWN
        }
    }
}

/**
 * Android mirror of the iOS `SavedSession` (`apps/ios/Sources/Models/SavedSessionsStore.swift`).
 *
 * One row in the persisted archived-session index. Keeps just enough to
 * render a History row read-only after the live session has ended and
 * dropped out of the broker's active list: a friendly summary (the first
 * user message), the agent, the originating server, and the last time we
 * saw it. The transcript itself is NOT stored here — it stays fetchable
 * from the broker's `archived-sessions/<id>/conversation.jsonl` via
 * [SessionStore.fetchConversation].
 */
data class SavedSession(
    val id: String,
    val serverId: String,
    val agent: String,
    val cwd: String?,
    val firstSeen: String,
    val lastSeen: String,
    val messageCount: Int,
    val summary: String,
    val status: SavedSessionStatus,
) {
    /** Compound identity matching the iOS/Rust `compound_key`. */
    val compoundId: String get() = "$serverId::$id"
}

/**
 * Pure reducers for the archived-session index — pulled out so JUnit can
 * exercise the archive-vs-permanent-delete contract without a live
 * ViewModel or EncryptedSharedPreferences (which need the AndroidKeyStore
 * at runtime). Mirror of the iOS `SavedSessionsStore.upsert` / `remove`
 * algorithms and the Rust `core/src/saved/mod.rs` rules.
 */
internal object SavedSessionsReducer {

    /** UTF-8 budget for the summary line — matches the iOS/Rust 100-char cap. */
    const val SUMMARY_MAX_CHARS = 100

    /** Upper bound on retained rows so the index can't grow unbounded. */
    const val INDEX_CAP = 500

    /**
     * Fold a live [ProjectSession] + [SessionStatus] snapshot into the
     * archived index. Idempotent — the same input twice returns an equal
     * list. `firstSeen` is sticky, `lastSeen` advances, the first non-empty
     * summary wins, and once a row is `EXITED` it stays that way. A
     * tombstoned id ([deleted]) is never (re-)added: a permanently-deleted
     * session must stay gone even though the broker can keep reporting it
     * (#199). Mirror of iOS `SavedSessionsStore.upsert`.
     */
    fun upsert(
        current: List<SavedSession>,
        session: ProjectSession,
        serverId: String,
        status: SessionStatus?,
        firstUserMessage: String?,
        messageCount: Int,
        isExited: Boolean,
        deleted: Set<String>,
        nowIso: String,
    ): List<SavedSession> {
        if (session.id in deleted) return current
        val now = status?.lastActivityAt
            ?: status?.startedAt
            ?: session.lastActivityAt
            ?: session.startedAt
            ?: nowIso
        val summary = truncateSummary(firstUserMessage ?: "")
        val nextStatus = if (isExited) SavedSessionStatus.EXITED else SavedSessionStatus.LIVE
        val agent = session.assistant
        val cwd = status?.cwd ?: session.cwd

        val idx = current.indexOfFirst { it.id == session.id && it.serverId == serverId }
        if (idx >= 0) {
            val row = current[idx]
            val merged = row.copy(
                lastSeen = maxOf(row.lastSeen, now),
                messageCount = maxOf(row.messageCount, messageCount.coerceAtLeast(0)),
                agent = agent,
                cwd = cwd ?: row.cwd,
                summary = if (row.summary.isEmpty() && summary.isNotEmpty()) summary else row.summary,
                status = mergeStatus(row.status, nextStatus),
            )
            if (merged == row) return current
            return current.toMutableList().also { it[idx] = merged }
        }
        val row = SavedSession(
            id = session.id,
            serverId = serverId,
            agent = agent,
            cwd = cwd,
            firstSeen = now,
            lastSeen = now,
            messageCount = messageCount.coerceAtLeast(0),
            summary = summary,
            status = nextStatus,
        )
        return capNewestFirst(current + row)
    }

    /**
     * Remove every row whose session id equals [id] from the index. Used
     * by permanent delete; archiving does NOT call this (the row stays).
     * Returns the input unchanged when nothing matched. Mirror of iOS
     * `SavedSessionsStore.remove(id:)`'s row-removal half (the tombstone
     * half lives in [SessionStore.tombstone]).
     */
    fun remove(current: List<SavedSession>, id: String): List<SavedSession> {
        val next = current.filterNot { it.id == id }
        return if (next.size == current.size) current else next
    }

    /**
     * Latest-first slice clamped to [limit], excluding tombstoned ids
     * (belt-and-braces — a permanently-deleted row is already removed, but
     * filter here too so a race can never leak one into History). Ties
     * broken by id for deterministic snapshot tests. Mirror of iOS
     * `SavedSessionsStore.recent`.
     */
    fun recent(
        current: List<SavedSession>,
        deleted: Set<String>,
        limit: Int = 200,
    ): List<SavedSession> {
        val sorted = current
            .filterNot { it.id in deleted }
            .sortedWith(compareByDescending<SavedSession> { it.lastSeen }.thenBy { it.id })
        return if (sorted.size <= limit) sorted else sorted.take(limit)
    }

    /** Exited is terminal (Unknown ⊏ Live ⊏ Exited). Mirror of iOS `mergeStatus`. */
    fun mergeStatus(existing: SavedSessionStatus, next: SavedSessionStatus): SavedSessionStatus {
        if (existing == SavedSessionStatus.EXITED || next == SavedSessionStatus.EXITED) {
            return SavedSessionStatus.EXITED
        }
        if (existing == SavedSessionStatus.LIVE || next == SavedSessionStatus.LIVE) {
            return SavedSessionStatus.LIVE
        }
        return SavedSessionStatus.UNKNOWN
    }

    /**
     * First-line, trimmed, ellipsized to [SUMMARY_MAX_CHARS]. Mirror of
     * iOS `truncateSummary` / Rust `truncate_summary`.
     */
    fun truncateSummary(text: String): String {
        val firstLine = text.lineSequence().firstOrNull().orEmpty().trim()
        if (firstLine.length <= SUMMARY_MAX_CHARS) return firstLine
        return firstLine.take(SUMMARY_MAX_CHARS - 1) + "…"
    }

    /** Keep the newest [INDEX_CAP] rows by `lastSeen` so the index can't grow forever. */
    private fun capNewestFirst(rows: List<SavedSession>): List<SavedSession> {
        if (rows.size <= INDEX_CAP) return rows
        return rows.sortedByDescending { it.lastSeen }.take(INDEX_CAP)
    }

    // MARK: - JSON persistence

    /**
     * Encode to the same envelope shape iOS/Rust use:
     * `{"sessions": {"<server>::<id>": {...}, ...}}` so the file stays
     * cross-platform round-trippable.
     */
    fun encode(rows: List<SavedSession>): String {
        val sessions = JSONObject()
        rows.forEach { r ->
            sessions.put(
                r.compoundId,
                JSONObject().apply {
                    put("id", r.id)
                    put("server_id", r.serverId)
                    put("agent", r.agent)
                    r.cwd?.let { put("cwd", it) }
                    put("first_seen", r.firstSeen)
                    put("last_seen", r.lastSeen)
                    put("message_count", r.messageCount)
                    put("summary", r.summary)
                    put("status", r.status.raw)
                },
            )
        }
        return JSONObject().apply { put("sessions", sessions) }.toString()
    }

    fun decode(raw: String?): List<SavedSession> {
        if (raw.isNullOrBlank()) return emptyList()
        return runCatching {
            val obj = JSONObject(raw)
            val sessions = obj.optJSONObject("sessions") ?: return@runCatching emptyList<SavedSession>()
            buildList {
                sessions.keys().forEach { key ->
                    val o = sessions.optJSONObject(key) ?: return@forEach
                    add(
                        SavedSession(
                            id = o.optString("id", ""),
                            serverId = o.optString("server_id", ""),
                            agent = o.optString("agent", ""),
                            cwd = o.optString("cwd", "").takeIf { it.isNotEmpty() },
                            firstSeen = o.optString("first_seen", ""),
                            lastSeen = o.optString("last_seen", ""),
                            messageCount = o.optInt("message_count", 0),
                            summary = o.optString("summary", ""),
                            status = SavedSessionStatus.fromRaw(o.optString("status", "unknown")),
                        ),
                    )
                }
            }.filter { it.id.isNotEmpty() }
        }.getOrDefault(emptyList())
    }
}
